lib/msf/core/exploit/remote/ms_lsarpc.rb

```diff
@@ -0,0 +1,136 @@
+###
+#
+# This mixin provides methods to open, and close policy handles, and to query policy info on the remote SMB server.
+#
+# -*- coding: binary -*-
+
+module Msf
+
+  module Exploit::Remote::MsLsarpc
+
+    include Msf::Exploit::Remote::MsIpc
+
+    class MsLsarpcError < StandardError; end
+    class MsLsarpcConnectionError < MsLsarpcError; end
+    class MsLsarpcAuthenticationError < MsLsarpcError; end
+    class MsLsarpcUnexpectedReplyError < MsLsarpcError; end
+
+    LSA_UUID        = '12345778-1234-abcd-ef00-0123456789ab'.freeze
+    LSA_VERS        = '0.0'.freeze
+    LSARPC_ENDPOINT = RubySMB::Dcerpc::Lsarpc.freeze
+
+    # The currently connected LSARPC pipe
+    attr_reader :lsarpc_pipe
+
+    def map_security_principal_to_string(security_principal)
+      case security_principal
+      when 1
+        'User'
+      when 2
+        'Group'
+      when 3
+        'Domain'
+      when 4
+        'Alias'
+      when 5
+        'Well-Known Group'
+      when 6
+        'Deleted Account'
+      when 7
+        'Invalid'
+      when 8
+        'Unknown'
+      when '9'
+        'Computer'
+      when 10
+        'Label'
+      else
+        'Unknown - Not a valid Security Principal'
+      end
+    end
+
+    def open_policy_2(impersonation_level, security_context_tracking_mode, access_mask)
+      object_attributes = LSARPC_ENDPOINT::LsaprObjectAttributes.new(
+        {
+          len: 24,
+          root_directory: nil,
+          object_name: nil,
+          attributes: 0,
+          security_descriptor: nil,
+          security_quality_of_service: {
+            len: 12,
+            impersonation_level: impersonation_level,
+            security_context_tracking_mode: security_context_tracking_mode,
+            effective_only: 0
+          }
+        }
+      )
+
+      self.lsarpc_pipe.lsar_open_policy_2(
+        system_name: simple.peerhost,
+        object_attributes: object_attributes,
```

Comment:
Once the [len](https://github.com/rapid7/ruby_smb/pull/267/files#r1595737180) change is in place, I believe you could simplify this substantially to just

```ruby
        object_attributes: {
          security_quality_of_service: {
            impersonation_level: impersonation_level,
            security_context_tracking_mode: security_context_tracking_mode
          }
        }
```

Other than being less code, it'd also be fewer names to change if that were required in the future. I'm pretty sure the strings will all default to nil and the numerics to 0.